MQE: Add support for changes function

lamida opened the pull request:
<!--  Thanks for sending a pull request!  Before submitting:

1. Read our CONTRIBUTING.md guide
2. Rebase your PR if it gets out of sync with main
-->

#### What this PR does

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated.
- n.a Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- n.a [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

jhesketh commented on pkg/streamingpromql/operators/functions/range_vectors.go:
```diff
@@ -369,3 +369,54 @@ func avgHistograms(head, tail []promql.HPoint) (*histogram.FloatHistogram, error
+	haveFloats := len(fHead) > 0 || len(fTail) > 0
+
+	if !haveFloats {
+		// PromQL engine doesn't support histogram for `changes` function yet,
```
(nit)
```suggestion
		// Prometheus' engine doesn't support histogram for `changes` function yet,
```

jhesketh commented on pkg/streamingpromql/operators/functions/range_vectors.go:
```diff
@@ -369,3 +369,54 @@ func avgHistograms(head, tail []promql.HPoint) (*histogram.FloatHistogram, error
+	return changesFloats(fHead, fTail), true, nil, nil
+}
+
+func changesFloats(head, tail []promql.FPoint) float64 {
```
(nit) probably doesn't need a separate function and can go inside changes above

jhesketh commented on pkg/streamingpromql/operators/functions/range_vectors.go:
```diff
@@ -369,3 +369,54 @@ func avgHistograms(head, tail []promql.HPoint) (*histogram.FloatHistogram, error
+
+	if len(tail) > 0 {
+		prev := tail[0].F
+		for _, sample := range tail[1:] {
```
rather than repeating this logic, set up a function you can pass head and tail into. Have a look at some other functions with `accumulate`

charleskorn commented on pkg/streamingpromql/operators/functions/range_vectors.go:
```diff
@@ -369,3 +369,54 @@ func avgHistograms(head, tail []promql.HPoint) (*histogram.FloatHistogram, error
+	}
+
+	if len(tail) > 0 {
+		prev := tail[0].F
```
This isn't quite right - if the last sample in `head` and the first sample in `tail` are different, we won't increment `changes`, but we should.

Perhaps we could get the first point in the ring buffer by calling `step.Floats.First()`, then iterate through the points?

lamida replied:
Updated in the latest commit

jhesketh commented on pkg/streamingpromql/operators/functions/range_vectors.go:
```diff
@@ -369,3 +369,54 @@ func avgHistograms(head, tail []promql.HPoint) (*histogram.FloatHistogram, error
+	}
+
+	if len(head) > 0 {
+		prev := head[0].F
```
I'm not sure this is correct as it is a ring buffer you'll probably need to look at the last element of tail. Might need to add a test of this boundary

lamida replied:
Updated in the latest commit

jhesketh reviewed (commented):
Nice work, thanks for that!

I think it's also worth putting this function through `TestCompareVariousMixedMetricsVectorSelectors` to test some edge cases (just add it to line 2310).

charleskorn commented on pkg/streamingpromql/operators/functions/range_vectors.go:
```diff
@@ -369,3 +369,54 @@ func avgHistograms(head, tail []promql.HPoint) (*histogram.FloatHistogram, error
+	haveFloats := len(fHead) > 0 || len(fTail) > 0
+
+	if !haveFloats {
+		// PromQL engine doesn't support histogram for `changes` function yet,
```
[nit] We use the term "Prometheus' engine" to refer to it - "PromQL engine" is ambiguous because MQE is a PromQL engine too.
```suggestion
		// Prometheus' engine doesn't support histogram for `changes` function yet,
```

Might also be good to add a link to the Prometheus' issue tracking adding support for native histograms in `changes`, if there is one.

lamida replied:
I tried to find but it seems there are no issue tracking for that.

jhesketh replied:
Charles is likely referring to this one: https://github.com/prometheus/prometheus/issues/13934

lamida replied:
That issue didn't mention changes. This is the comment that mentioned the need to add support ([code link](https://github.com/prometheus/prometheus/blob/7c7116fea8343795cae6da42960cacd0207a2af8/promql/functions.go#L1444)).

lamida commented:
@jhesketh @charleskorn I have addressed all the comments. These are changes from the previous reviews:
- Handle ring-buffer wrap around.
- Moved the function logic from `range_vector.go` to `fpoint_ring_buffer.go` so that I can test ring-buffer wrap around easier. Let me know if anyone have concern.

jhesketh commented on pkg/streamingpromql/types/fpoint_ring_buffer.go:
```diff
@@ -237,6 +239,49 @@ func (b *FPointRingBuffer) AnyAtOrBefore(maxT int64) bool {
 
+// ChangesAtOrBefore returns the number of changes between subsequent points in this ring buffer from the first
+// point to the point with timestamp less than or equal to maxT.
+func (b *FPointRingBuffer) ChangesAtOrBefore(maxT int64) (float64, bool) {
```
I think this is better fit into functions/range_vectors in terms of consistency. Can we not create a test in the functions package and pass it a constructed `types.RangeVectorStepData`?

lamida replied:
> I think this is better fit into functions/range_vectors in terms of consistency

Actually some functions are implemented in `fpoint_ring_buffer.go` too like `CountAtOrBefore`, `LastAtOrBefore` and `AnyAtOrBefore`.

> Can we not create a test in the functions package and pass it a constructed types.RangeVectorStepData?
It is not possible to create the buffer to simulate wrap around because `points` and other fields are private. I prefer not to make them public just for testing.

Other than that, if we put the logic in `range_vectors.go` we are exposing the internal representation of the ring buffer in `range_vector.go`. Like on how iterating head and tail which I believe should be internal `fpoint_ring_buffer.go`. 

At the end I don't have any strong opinion and I am fine if there is a better argument.

lamida replied:
> Can we not create a test in the functions package and pass it a constructed types.RangeVectorStepData?
It is not possible to create the buffer to simulate wrap around because the fields are private.

charleskorn replied:
I would favour this not being on the ring buffer type - `ChangesAtOrBefore` can efficiently be implemented by using `UnsafePoints`, whereas `CountAtOrBefore`, `LastAtOrBefore` and `AnyAtOrBefore` don't need the slices produced by `UnsafePoints` and so benefit from being implemented directly on the ring buffer type.

lamida replied:
I will moved back to `range_vectors.go` then. I will figure out how the best to test that.

jhesketh commented on pkg/streamingpromql/types/fpoint_ring_buffer.go:
```diff
@@ -237,6 +239,49 @@ func (b *FPointRingBuffer) AnyAtOrBefore(maxT int64) bool {
+		}
+	}
+
+	// The points buffer is wrapped around, therefore we need to check changes starting from the buffer's head
+	// and then continue to the tail.
+	if len(fHead) > 0 && len(fTail) > 0 {
+		// We don't need to compare the first point with itself, hence start from fHead[1:]
+		accumulate(fHead[1:])
+		accumulate(fTail)
+	}
+
+	if len(fHead) > 0 && len(fTail) == 0 {
+		// We don't need to compare the first point with itself, hence start from fHead[1:]
+		accumulate(fHead[1:])
+	}
```
This is consistent with how we use accumulate elsewhere. I don't think there is much overhead in range when the len is 0 vs using an if.
```suggestion
	// The points buffer is wrapped around, therefore we need to check changes starting from the buffer's head
	// and then continue to the tail.
	// We don't need to compare the first point with itself, hence start from fHead[1:]
	accumulate(fHead[1:])
	accumulate(fTail)
```

lamida replied:
You are right. We can simplify that. Updated in a7d04746f58fc416884ba3fbeb1f3536418eabfb

jhesketh commented on pkg/streamingpromql/types/fpoint_ring_buffer_test.go:
```diff
@@ -0,0 +1,155 @@
+					{T: 9, F: 99},
+					{T: 1, F: 1},
+					{T: 3, F: 2},
+					{T: 5, F: 3},
```
Would be good to have a test where the value doesn't change at the wrap around

lamida replied:
Added in af5c86715873ccde85c07e5064897864dc473be6

charleskorn commented on pkg/streamingpromql/operators/functions/range_vectors.go:
```diff
@@ -369,3 +369,43 @@ func avgHistograms(head, tail []promql.HPoint) (*histogram.FloatHistogram, error
+		}
+	}
+
+	accumulate(fHead)
```
Couldn't we skip the first point? Otherwise we'll just be comparing it to itself.

```suggestion
	accumulate(fHead[1:])
```

jhesketh replied:
Yes, I think this was already done in the previous revision?

charleskorn reviewed (approved):
LGTM modulo suggestion below

jhesketh reviewed (commented):
lgtm too, but agree with Charles. Thanks Jon!
